### DISCUSSION TITLE
PublicKeyCredentialCreationOptions to json

### DISCUSSION THREAD

**Discussion by santhoshit8** [webauthn4j/webauthn4j#692] [General]
Hi,

is there any utility class to convert PublicKeyCredentialCreationOptions to json , so that it can be sent to webauthn js api

Thanks,
santhosh

**Comment by ynojima** [comment]
Since there wasn't standard serialization format for `PublicKeyCredentialCreationOptions`, we leave the task to library users.
(For example, in which format is the byte array encoded? HexString? Base64String? Base64UrlString?)

You may define this kind Jackson Minin, and and use Jackson serializer with them to serialize the object.
https://github.com/webauthn4j/webauthn4j-spring-security/blob/9ed65b74479b814d18a1da9b9544174a792f7b5e/webauthn4j-spring-security-core/src/main/java/com/webauthn4j/springframework/security/endpoint/PublicKeyCredentialDescriptorMixin.java#L29-L35
https://github.com/webauthn4j/webauthn4j-spring-security/blob/9ed65b74479b814d18a1da9b9544174a792f7b5e/webauthn4j-spring-security-core/src/main/java/com/webauthn4j/springframework/security/endpoint/PublicKeyCredentialUserEntityMixin.java#L27-L31

**Reply by santhoshit8** [reply]
can we use jsonConverter ynojima?

```
`PublicKeyCredentialCreationOptions publicKeyCredentialCreationOptions = new PublicKeyCredentialCreationOptions(.....);
ObjectConverter objectConverter = new ObjectConverter();

JsonConverter jsonConverter = objectConverter.getJsonConverter();
String registrationRequest  = JsonConverter.writeValueAsString(publicKeyCredentialCreationOptions);`
```

**Reply by ynojima** [reply]
`JsonConverter` is a wrapper for Jackson serializer, but byte array serialization format is not specified.
When not specified, Jackson serializer uses Base64 encoding by default, but is it OK for you?
If you want to use Base64Url encoding, define mixins like this:
https://github.com/webauthn4j/webauthn4j-spring-security/blob/9ed65b74479b814d18a1da9b9544174a792f7b5e/webauthn4j-spring-security-core/src/main/java/com/webauthn4j/springframework/security/endpoint/PublicKeyCredentialDescriptorMixin.java#L29-L35
https://github.com/webauthn4j/webauthn4j-spring-security/blob/9ed65b74479b814d18a1da9b9544174a792f7b5e/webauthn4j-spring-security-core/src/main/java/com/webauthn4j/springframework/security/endpoint/PublicKeyCredentialUserEntityMixin.java#L27-L31
, attach it to the original class by defining Jackson Module class:
https://github.com/webauthn4j/webauthn4j-spring-security/blob/9ed65b74479b814d18a1da9b9544174a792f7b5e/webauthn4j-spring-security-core/src/main/java/com/webauthn4j/springframework/security/converter/jackson/WebAuthn4JSpringSecurityJSONModule.java
and create a ObjectConverter object by injecting the defined Jackson Module class:
https://github.com/webauthn4j/webauthn4j-spring-security/blob/a17d579f19a9f24dcbbbd8e942ab350b49c7ee7a/samples/spa/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/app/config/WebSecurityBeanConfig.java#L116-L120

**Reply by santhoshit8** [reply]
Thanks for info ynojima!

**Comment by santhoshit8** [comment]
Hi @ynojima 

I am using default challenge
`Challenge challenge = new DefaultChallenge();`

Converting PublicKeyCredentialCreationOptions to json using JsonConverter
`String registrationRequest  = JsonConverter.writeValueAsString(publicKeyCredentialCreationOptions);`

After converting, challenge include base64url character sometimes like as follow
`ZRb1IKFiSyyTkwOgt_ZMOQ`

if it is were base64 conversion, it should not contain "_".. right?

Any thought on this?

Thanks

**Reply by ynojima** [reply]
`ChallengeSerializer`, which serializes a `Challenge` into a base64url string, is included in `WebAuthnJSONModule`. This is why the `Challenge` is converted into base64url by default, but I cannot recall why I included it.... 🧐
https://github.com/webauthn4j/webauthn4j/blob/bc35f72c85e971244b50eebd73d9201f362fa7e1/webauthn4j-core/src/main/java/com/webauthn4j/converter/jackson/WebAuthnJSONModule.java#L58

**Reply by santhoshit8** [reply]
Thanks for info @ynojima
